### ISSUE TITLE
Add option to save layouts

### ISSUE THREAD

**Issue by SimonBrandner** [hyprwm/Hyprland#1426]
Add option to save layouts similarly to what [i3wm allows you to do it](https://i3wm.org/docs/layout-saving.html)

**Comment by yogeshkumar98103** [comment]
It'll be very useful to have an option to automatically restore last layout on reboot.

**Comment by magicBeast** [comment]
When doing development I often have the same four or more windows open, in the same layout. It would be useful to be able to save a session and quickly restore any required layout for a given task or set of tasks. Obviously this can be hacked together via scripts, but it does seem like natural extension of functionality for a tiling compositor that handles window layout.

**Comment by witnes0417** [comment]
yes this would be very useful, even though tmux complement in its way, but why need to configure another window navigating feature in it?

**Comment by memchr** [comment]
How would some applications that open multiple windows, so that they all have the same PID, be distinguished? They may also have the same criteria.

**Comment by ScuttleSE** [comment]
I just assumed this was available when I installed Hyprland and I must say...disappointed that it is not available, and a pull request to actually add it was rejected...

**Comment by EnderSpy29** [comment]
yo i'd love this, having all my windows opening with certain sizes in certain positions in certain workspaces would be amazing. i'd be great if we could either take a snapshot and bind that to a key or something. or have hyprland load the last used layout on startup

**Comment by tonysvo** [comment]
agreed, I want to have few workspace pre-setup every time I login

**Comment by izmyname** [comment]
As a workaround, you can autostart desired programs. Here is an example from my config.

` exec-once=[workspace 1 silent] $browser & sleep 4; $clock & sleep 1; $filemanager & $telegram &`

**Comment by izmyname** [comment]
Also, I see several pull requests, relevant to this issue, were declined, so I guess this function isn't going to be implemented. Should the issue be closed?

**Comment by EnderSpy29** [comment]
> As a workaround, you can autostart desired programs. Here is an example from my config.
> 
> ` exec-once=[workspace 1 silent] $browser & sleep 4; $clock & sleep 1; $filemanager & $telegram &`

i have programs executed on startup. what we want is the ability to set all applications to load in a predefined location, size and layout

**Comment by zleeepy** [comment]
i need this

**Comment by Fork3r** [comment]
oh god this is musthave feature

**Comment by Cool-Game-Dev** [comment]
I want to save my screenshot layout because its a bit annoying to set up

**Comment by agluck91** [comment]
I would love to see this. I am trying to figure out have to program it with hyprctl commands. I think it is possible. If I get it working I will post a link in this thread. Work around until they can get it implemented in the core.

**Comment by vaxerski** [comment]
please use the :+1: reaction instead of useless "me too" comments, thanks.

**Comment by Natr1x** [comment]
I made a sort of "proof of concept" for something which can be used with the dwindle layout #9054 

It doesn't save or restore any layouts by itself but instead aims to add the necessary features to make it possible to do with hyprctl.

**Comment by prijatelj** [comment]
I made a python script that loads the prior windows into their prior workspace from the json that results from `hyprctl -j clients`.

https://github.com/prijatelj/dots/blob/sys76/.config/hypr/load_layout

I primarily need this for Firefox where i have it restore the prior open windows restarting Firefox. So it works for such apps, and currently only places the windows in their original workspaces without reorganizing or resizing them.

Any unassigned windows remain in a temporary workspace for the user to manually handle.

I use the dwindle layout and have not tested for master layout, though given the dispatchers are basically just `exec` and `movetoworkspace`, I suspect it works for both.

Regarding this https://github.com/hyprwm/Hyprland/pull/9054#issuecomment-2602069170 , the layout jsons seem to have enough information saved per workspace such that you *can* recreate the exact layout given the `"at" : [x, y]; "size": [width, height]` info and if you know the monitor resolution, e.g., from `xrandr`, as long as the layout type (dwindle/master) is known and workspace movement & resizing is deterministic.   May have to tell it to ignore cursor placement for some commands.

This was not straight forward, so I didn't get to it yet.

My ideal setup will have that + handle multiple call per window commands like a terminal + a listener to layout changes to save it automatically.  That's for future free time though.

Feel free to use this script and extend!

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!
